src/audio/codec_adapter/codec/generic.c

```diff
@@ -277,7 +277,11 @@ int codec_process(struct comp_dev *dev)
 
 	comp_dbg(dev, "codec_process() done");
 out:
-	codec->state = CODEC_IDLE;
+	/* Was there reset request? */
```

Comment:
so the codec causes a reset by changing its own state?

Reply:
We get reset request from external world however if we perform processing at the time there is no safe, scalable way to terminate processing and do the reset. So instead of doing it right away we register the request and actually reset when idle.

Reply:
@mrajwa can you elaborate a bit more, whats the real world use case when the codec has to reset itself.

Reply:
@lgirdwood request comes while codec is processing - we don't want to tinker with its memory, so instead we change the state to resetting, check for that at the end of processing and finally reset itself. Does it make sense now?

Reply:
ok, so IIUC we have the following sequence
1) codec is processing
2) IPC trigger(STOP)
3) codec completes processing
4) codec then handles the STOP ?

Reply:
Yes, that's assumed flow.

Reply:
Given the component lifecycle graph, is there a possibility that we could not get another copy call after we get the reset call? I am just wondering how we make sure that copy is called again after reset as it seems opposite to the graph.

Reply:
Can you link to the graph you are talking about?
Also, not this is not about getting into another copy, its about interrupting the current one.

Reply:
@cujomalainey @lgirdwood I have reviewed the current IPC/IDC implementation again and I think we are already safe from this issue. Previously I was thinking about the scenario when the processing on the second core takes longer than the processing on the main one (can happen) and IPC reset handler will interrupt our processing on the slave core. But this won't happen because the request for the slave core is handled in the edf task which comes into play after our pipeline task. That said I am going to drop that one commit which verifies the state of the codec during the reset method.

Reply:
FTR lifecycle is here https://thesofproject.github.io/latest/developer_guides/firmware/components/component-overview.html#handle-the-component-device-state